# In acest notebook vor fi descrise cateva concepte de baza ale tensorilor utilizand TensorFlow

# Content
## * Introducere in tensor
## * Obtinerea informatiei despre tensor
## * Manipularea tensorilor (opratii cu tensorii)
## * Tensorii si NumPy
## * Utilizarea @tf.function (cresterea vitezei functiilor Python)
## * Utilizarea GPU in TensorFlow (sau TPU)
## * Exercitii de exersare

# 1. Introducere in Tensor

## 1.1. Importul bibiotecilor

In [227]:
# Importul TensorFlow
import tensorflow as tf

# Import NumPy
import numpy as np

In [228]:
# verificarea versiunii TensorFlow
print(tf.__version__)

2.4.1


## 1.2. Crearea unui tensor

### 1.2.1. Crearea unui tenso cu ajutorul tf.constant()

In [229]:
# Crearea unui tensor scalar
scalar = tf.constant(9)
scalar

<tf.Tensor: shape=(), dtype=int32, numpy=9>

In [230]:
# Verificarea dimensiunilor cu ajutorul atributului ndim
scalar.ndim

0

In [231]:
# Crearea unui tensor vector
vector = tf.constant([10, 9])
vector

<tf.Tensor: shape=(2,), dtype=int32, numpy=array([10,  9], dtype=int32)>

In [232]:
# verificarea dimensiunii
vector.ndim

1

In [233]:
# Crearea unui tensor matrice
matrix = tf.constant([[10, 7],
                     [7, 3]])
matrix

<tf.Tensor: shape=(2, 2), dtype=int32, numpy=
array([[10,  7],
       [ 7,  3]], dtype=int32)>

In [234]:
# verificarea dimensiunii
matrix.ndim

2

In [235]:
# Crearea unui tensor matrice cu specificarea tipului de date
matrix_1 = tf.constant([[10., 7.],
                        [7., 3.],
                        [3., 6.]], dtype=tf.float16)
matrix_1

<tf.Tensor: shape=(3, 2), dtype=float16, numpy=
array([[10.,  7.],
       [ 7.,  3.],
       [ 3.,  6.]], dtype=float16)>

In [236]:
# verificarea dimensiunii
matrix_1.ndim

2

In [237]:
# Crearea unui tensor 3-dimensional
tensor = tf.constant([[[10, 7],
                        [7, 3],
                        [5, 9]],
                      [[3, 6],
                        [3, 7],
                        [2, 1]]])
tensor

<tf.Tensor: shape=(2, 3, 2), dtype=int32, numpy=
array([[[10,  7],
        [ 7,  3],
        [ 5,  9]],

       [[ 3,  6],
        [ 3,  7],
        [ 2,  1]]], dtype=int32)>

In [238]:
# verificarea dimensiunii
tensor.ndim

3

### 1.2.2. Crearea unui tensor cu ajutorul tf.Variable()

In [239]:
# crearea unui vector variabil
tensor_var=tf.Variable([10,9])
tensor_var

<tf.Variable 'Variable:0' shape=(2,) dtype=int32, numpy=array([10,  9], dtype=int32)>

In [240]:
# Citirea unui element din tensor_var cu ajutorul indexului
tensor_var[0]

<tf.Tensor: shape=(), dtype=int32, numpy=10>

In [241]:
# Modificarea unui element in tensor_var utilizand functia assign()
tensor_var[0].assign(7)
tensor_var


<tf.Variable 'Variable:0' shape=(2,) dtype=int32, numpy=array([7, 9], dtype=int32)>

### 1.2.3. Crearea unor tensori aliatori

In [242]:
# Crearea unui tensor aliator cu distributie uniforma
random_0 = tf.random.Generator.from_seed(9)
random_0 = tf.random.uniform(shape=(3,2))
random_0

<tf.Tensor: shape=(3, 2), dtype=float32, numpy=
array([[0.46696687, 0.7581432 ],
       [0.6996453 , 0.2605592 ],
       [0.5604588 , 0.26713252]], dtype=float32)>

In [243]:
# Crearea unui tensor aliator cu distributie normata
random_1 = tf.random.Generator.from_seed(9) 
random_1 = random_1.normal(shape=(3, 2))
random_1

<tf.Tensor: shape=(3, 2), dtype=float32, numpy=
array([[-0.6039789 , -0.1766927 ],
       [ 0.04221033,  0.29037967],
       [-0.29604465, -0.21134207]], dtype=float32)>

In [244]:
# Crearea unui alt tensor aliator cu distributie normata
random_2 = tf.random.Generator.from_seed(9) 
random_2 = random_2.normal(shape=(3, 2))
random_2

<tf.Tensor: shape=(3, 2), dtype=float32, numpy=
array([[-0.6039789 , -0.1766927 ],
       [ 0.04221033,  0.29037967],
       [-0.29604465, -0.21134207]], dtype=float32)>

In [245]:
# Compararea valorilor celor doi tensori cu distribuitie normata
random_1==random_2

<tf.Tensor: shape=(3, 2), dtype=bool, numpy=
array([[ True,  True],
       [ True,  True],
       [ True,  True]])>

### 1.2.4. Amestecarea ordinii elementelor intr-un tensor

In [246]:
# Crearea unui tensor
tensor_init = tf.constant([[10, 7],
                            [3, 4],
                            [2, 5]])
tensor_init

<tf.Tensor: shape=(3, 2), dtype=int32, numpy=
array([[10,  7],
       [ 3,  4],
       [ 2,  5]], dtype=int32)>

In [247]:
# Amestecarea tensor
tensor_mix_0 = tf.random.shuffle(tensor_init)
tensor_mix_0

<tf.Tensor: shape=(3, 2), dtype=int32, numpy=
array([[ 2,  5],
       [ 3,  4],
       [10,  7]], dtype=int32)>

In [248]:
# Amestecarea repetata tensor
tensor_mix_1 = tf.random.shuffle(tensor_init)
tensor_mix_1

<tf.Tensor: shape=(3, 2), dtype=int32, numpy=
array([[ 3,  4],
       [ 2,  5],
       [10,  7]], dtype=int32)>

In [249]:
# Compararea valorilor celor doi tensori
tensor_mix_0==tensor_mix_1

<tf.Tensor: shape=(3, 2), dtype=bool, numpy=
array([[False, False],
       [False, False],
       [ True,  True]])>

In [250]:
# Amestecarea tensor cu setarea parametrului seed
tf.random.set_seed(9)
tensor_mix_2 = tf.random.shuffle(tensor_init, seed=9)
tensor_mix_2

<tf.Tensor: shape=(3, 2), dtype=int32, numpy=
array([[10,  7],
       [ 3,  4],
       [ 2,  5]], dtype=int32)>

In [251]:
# Amestecarea repretata a tensorului cu setarea parametrului seed
tf.random.set_seed(9) # seed nivel global
tensor_mix_3 = tf.random.shuffle(tensor_init, seed=9) # seed nivel operational
tensor_mix_3

<tf.Tensor: shape=(3, 2), dtype=int32, numpy=
array([[10,  7],
       [ 3,  4],
       [ 2,  5]], dtype=int32)>

In [252]:
# Compararea valorilor ultimilor doi tensori
tensor_mix_2==tensor_mix_3

<tf.Tensor: shape=(3, 2), dtype=bool, numpy=
array([[ True,  True],
       [ True,  True],
       [ True,  True]])>

### 1.2.5. Crearea unui tensor din NumPy arrays

In [253]:
# Crearea unui tensor ce contine doar elemete 1
tf.ones(shape=(5, 4))

<tf.Tensor: shape=(5, 4), dtype=float32, numpy=
array([[1., 1., 1., 1.],
       [1., 1., 1., 1.],
       [1., 1., 1., 1.],
       [1., 1., 1., 1.],
       [1., 1., 1., 1.]], dtype=float32)>

In [254]:
# Crearea unui tensor ce contine doar elemete 0
tf.zeros(shape=(4, 6))

<tf.Tensor: shape=(4, 6), dtype=float32, numpy=
array([[0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0.]], dtype=float32)>

In [255]:
# Crearea unui numpy array
array_A = np.arange(1, 25, dtype=np.int32)
array_A

array([ 1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16, 17,
       18, 19, 20, 21, 22, 23, 24], dtype=int32)

In [256]:
# Conversia numpy array intr-un tensor vector
tensor_A = tf.constant(array_A) 
tensor_A

<tf.Tensor: shape=(24,), dtype=int32, numpy=
array([ 1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16, 17,
       18, 19, 20, 21, 22, 23, 24], dtype=int32)>

In [257]:
# Verificarea dimensiunilor tensor_A
tensor_A.ndim 

1

In [258]:
# Conversia numpy array intr-un tensor 3-dimensional
Tensor_A = tf.constant(array_A, shape=(2, 3, 4)) 
Tensor_A

<tf.Tensor: shape=(2, 3, 4), dtype=int32, numpy=
array([[[ 1,  2,  3,  4],
        [ 5,  6,  7,  8],
        [ 9, 10, 11, 12]],

       [[13, 14, 15, 16],
        [17, 18, 19, 20],
        [21, 22, 23, 24]]], dtype=int32)>

In [259]:
# Verificarea dimensiunilor Tensor_A
Tensor_A.ndim 

3

Conform conventiei, denumirile vectorilor incep cu minuscula, iar a matricilor si a tensorilor n-dimensionali cu majuscula

Diferenta dintre NumPy arrays si tensor consta in faptul ca tensor poate fi excuta pe un GPU (viteza mai mare de executie a operatiilor)

# 2. Obtinerea informatiei despre tensor

## 2.1. Informatii despre dimensiuni/forma

Din cadrul unui tensor se poate extrage informatii despre acesta:


* Shape: Dimensiunea (numarul de elemente) a fiecarea dimensiuni ale tensorului.
* Rank: Numarul dimensiunii tensorului. Un scalar alre rank =0, un verctor rank=1, o matrice rank=2 iar un tensor are rank=n.
* Axis or Dimension: O dimensiune particulara a tensorului.
* Size: Numarul total de elemente intr-un tensor.

In [260]:
#Crearea unui tensor 
Tensor_0 = tf.zeros([2, 3, 5, 4])
Tensor_0 

<tf.Tensor: shape=(2, 3, 5, 4), dtype=float32, numpy=
array([[[[0., 0., 0., 0.],
         [0., 0., 0., 0.],
         [0., 0., 0., 0.],
         [0., 0., 0., 0.],
         [0., 0., 0., 0.]],

        [[0., 0., 0., 0.],
         [0., 0., 0., 0.],
         [0., 0., 0., 0.],
         [0., 0., 0., 0.],
         [0., 0., 0., 0.]],

        [[0., 0., 0., 0.],
         [0., 0., 0., 0.],
         [0., 0., 0., 0.],
         [0., 0., 0., 0.],
         [0., 0., 0., 0.]]],


       [[[0., 0., 0., 0.],
         [0., 0., 0., 0.],
         [0., 0., 0., 0.],
         [0., 0., 0., 0.],
         [0., 0., 0., 0.]],

        [[0., 0., 0., 0.],
         [0., 0., 0., 0.],
         [0., 0., 0., 0.],
         [0., 0., 0., 0.],
         [0., 0., 0., 0.]],

        [[0., 0., 0., 0.],
         [0., 0., 0., 0.],
         [0., 0., 0., 0.],
         [0., 0., 0., 0.],
         [0., 0., 0., 0.]]]], dtype=float32)>

In [261]:
# Obtinerea informatiei despre tensorul creat
Tensor_0.shape, Tensor_0.ndim, tf.size(Tensor_0)

(TensorShape([2, 3, 5, 4]), 4, <tf.Tensor: shape=(), dtype=int32, numpy=120>)

In [262]:
# Crearea unei functii de obtinere a diferitor atribute despre tensor
def tensor_atributes(Tensor):
  """
  Afiseaza diferite informatii despre tensor
  """
  print("Tipul de date a fiecarui element:", Tensor.dtype)
  print("Numarul de dimensiuni (rank):", Tensor.ndim)
  print("Forma tensorului:", Tensor.shape)
  print("Numarul de elemente pe axa 0 a tensorului:", Tensor.shape[0])
  print("numarul de elemente pe ultima axa a tensorului:", Tensor.shape[-1])
  print("Numarul total de elemente:", tf.size(Tensor).numpy()) # .numpy() converteste in NumPy array

In [263]:
# Obtinerea diferitor atribute despre Tensor_0
tensor_atributes(Tensor_0)

Tipul de date a fiecarui element: <dtype: 'float32'>
Numarul de dimensiuni (rank): 4
Forma tensorului: (2, 3, 5, 4)
Numarul de elemente pe axa 0 a tensorului: 2
numarul de elemente pe ultima axa a tensorului: 4
Numarul total de elemente: 120


## 2.2. Informatii despre elemente utilizand indexarea

In [264]:
# Obtinerea primelor 2 elemente ale fiecareai dimensiuni
Tensor_0[:2, :2, :2, :2]

<tf.Tensor: shape=(2, 2, 2, 2), dtype=float32, numpy=
array([[[[0., 0.],
         [0., 0.]],

        [[0., 0.],
         [0., 0.]]],


       [[[0., 0.],
         [0., 0.]],

        [[0., 0.],
         [0., 0.]]]], dtype=float32)>

In [265]:
# Obtinerea primelor elemente ale fiecarei dimensiuni, exceptand ultima care va contine toate elementele
Tensor_0[:1, :1, :1, :]

<tf.Tensor: shape=(1, 1, 1, 4), dtype=float32, numpy=array([[[[0., 0., 0., 0.]]]], dtype=float32)>

In [266]:
# Obtinerea primelor elemente ale fiecarei dimensiuni, exceptand penultima care va contine toate elementele
Tensor_0[:1, :1, :, :1]

<tf.Tensor: shape=(1, 1, 5, 1), dtype=float32, numpy=
array([[[[0.],
         [0.],
         [0.],
         [0.],
         [0.]]]], dtype=float32)>

## 2.3. Marirea numarului de dimensiunii ale tensorului

In [267]:
# Crearea unui tensor cu rank=2
Tensor_2 = tf.constant([[8, 9],
                        [4, 3]])
Tensor_2

<tf.Tensor: shape=(2, 2), dtype=int32, numpy=
array([[8, 9],
       [4, 3]], dtype=int32)>

In [268]:
# Obtinerea diferitor atribute despre Tensor_2
tensor_atributes(Tensor_2)

Tipul de date a fiecarui element: <dtype: 'int32'>
Numarul de dimensiuni (rank): 2
Forma tensorului: (2, 2)
Numarul de elemente pe axa 0 a tensorului: 2
numarul de elemente pe ultima axa a tensorului: 2
Numarul total de elemente: 4


### 2.3.1. Adaugarea unei dimensiuni utilizand `tf.newaxis`

In [269]:
# Crearea unui nou tensor din tensor_2 adaugand inca o dimensiune
Tensor_3 = Tensor_2[..., tf.newaxis] #  "..." insemnand toate dimensiunile anterioare
Tensor_3

<tf.Tensor: shape=(2, 2, 1), dtype=int32, numpy=
array([[[8],
        [9]],

       [[4],
        [3]]], dtype=int32)>

In [270]:
# Obtinerea diferitor atribute despre Tensor_3
tensor_atributes(Tensor_3)

Tipul de date a fiecarui element: <dtype: 'int32'>
Numarul de dimensiuni (rank): 3
Forma tensorului: (2, 2, 1)
Numarul de elemente pe axa 0 a tensorului: 2
numarul de elemente pe ultima axa a tensorului: 1
Numarul total de elemente: 4


### 2.3.2. Adaugarea unei dimensiuni utilizand `tf.expand_dims().`

In [271]:
# Crearea unui nou tensor din tensor_2 adaugand inca o dimensiune la final
Tensor_4 = tf.expand_dims(Tensor_2, axis=-1) # "-1" insemnand ultima axa
Tensor_4

<tf.Tensor: shape=(2, 2, 1), dtype=int32, numpy=
array([[[8],
        [9]],

       [[4],
        [3]]], dtype=int32)>

In [272]:
# Obtinerea diferitor atribute despre Tensor_4
tensor_atributes(Tensor_4)

Tipul de date a fiecarui element: <dtype: 'int32'>
Numarul de dimensiuni (rank): 3
Forma tensorului: (2, 2, 1)
Numarul de elemente pe axa 0 a tensorului: 2
numarul de elemente pe ultima axa a tensorului: 1
Numarul total de elemente: 4


In [273]:
# Crearea unui nou tensor din tensor_2 adaugand inca o dimensiune la inceput
Tensor_5 = tf.expand_dims(Tensor_2, axis=0) # "0" insemnand prima axa
Tensor_5

<tf.Tensor: shape=(1, 2, 2), dtype=int32, numpy=
array([[[8, 9],
        [4, 3]]], dtype=int32)>

In [274]:
# Obtinerea diferitor atribute despre Tensor_5
tensor_atributes(Tensor_5)

Tipul de date a fiecarui element: <dtype: 'int32'>
Numarul de dimensiuni (rank): 3
Forma tensorului: (1, 2, 2)
Numarul de elemente pe axa 0 a tensorului: 1
numarul de elemente pe ultima axa a tensorului: 2
Numarul total de elemente: 4


# 3. Manipularea tensorilor (operatii cu tensorii)

## 3.1. Operatii de baza

In [275]:
# Afisarea valorilor Tensor_2
Tensor_2

<tf.Tensor: shape=(2, 2), dtype=int32, numpy=
array([[8, 9],
       [4, 3]], dtype=int32)>

In [276]:
# Adunarea unui numar scalar fiecarui element al Tensor_2
Tensor_2+10


<tf.Tensor: shape=(2, 2), dtype=int32, numpy=
array([[18, 19],
       [14, 13]], dtype=int32)>

In [277]:
# Multiplicarea fiecarui element al Tensor_2
Tensor_2*2

<tf.Tensor: shape=(2, 2), dtype=int32, numpy=
array([[16, 18],
       [ 8,  6]], dtype=int32)>

In [278]:
# Scaderea unui numar scalar din fiecare element al Tensor_2
Tensor_2-10

<tf.Tensor: shape=(2, 2), dtype=int32, numpy=
array([[-2, -1],
       [-6, -7]], dtype=int32)>

In [279]:
# Multiplicarea fiecarui element al Tensor_2 utilizand functia in-build
tf.multiply(Tensor_2, 3)

<tf.Tensor: shape=(2, 2), dtype=int32, numpy=
array([[24, 27],
       [12,  9]], dtype=int32)>

In [280]:
# Adunarea unui numar scalar fiecarui element al Tensor_2 utilizand functia in-build
tf.add(Tensor_2, 9)

<tf.Tensor: shape=(2, 2), dtype=int32, numpy=
array([[17, 18],
       [13, 12]], dtype=int32)>

## 3.2. Inmultirea matricilor

Teoria inmultirii matricilor poate fi gasita pe adresele:

https://www.mathsisfun.com/algebra/matrix-multiplying.html

http://matrixmultiplication.xyz/

In [281]:
# Crearea unui tensor Tensor_0
Tensor_0 = tf.constant([[4, 7, 9],
                        [5, 2, 8]])
Tensor_0

<tf.Tensor: shape=(2, 3), dtype=int32, numpy=
array([[4, 7, 9],
       [5, 2, 8]], dtype=int32)>

In [282]:
# Crearea unui  alt tensor Tensor_1
Tensor_1 = tf.constant([[3, 2],
                        [6, 4],
                        [7, 9]])
Tensor_1

<tf.Tensor: shape=(3, 2), dtype=int32, numpy=
array([[3, 2],
       [6, 4],
       [7, 9]], dtype=int32)>

In [283]:
# Multiplicarea matricilor utilizand metoda tf.matmul()
tf.matmul(Tensor_0, Tensor_1)

<tf.Tensor: shape=(2, 2), dtype=int32, numpy=
array([[117, 117],
       [ 83,  90]], dtype=int32)>

In [284]:
# Multiplicarea matricilor utilizand operatorul Python @
Tensor_0 @ Tensor_1

<tf.Tensor: shape=(2, 2), dtype=int32, numpy=
array([[117, 117],
       [ 83,  90]], dtype=int32)>

In [285]:
# Multiplicarea matricilor utilizand tf.tensordot()
tf.tensordot(Tensor_0, Tensor_1, axes=1)

<tf.Tensor: shape=(2, 2), dtype=int32, numpy=
array([[117, 117],
       [ 83,  90]], dtype=int32)>

In [286]:
# Afisarea valorilor Tensor_2
Tensor_2

<tf.Tensor: shape=(2, 2), dtype=int32, numpy=
array([[8, 9],
       [4, 3]], dtype=int32)>

In [287]:
# Vizualizare formei matricilor Tensor_0 si Tensor_2
Tensor_0.shape, Tensor_2.shape

(TensorShape([2, 3]), TensorShape([2, 2]))

In [288]:
# Incercarea de a inmulti 2 matric a caror dimensiuni sunt incompatibile
#Tensor_0 @ Tensor_2

Reguli referitoare la dimensiunile matricilor ce vor fi inmultite:
1. Dmmensiunile interne trebuie sa coincida
2. Matricea rezultata va avea dimensiunile externe

               `(m × n) * (n × p) = (m × p)`


## 3.3. Redimensionare unui tensor

In [289]:
# Vizualizarea Tensor_0
Tensor_0

<tf.Tensor: shape=(2, 3), dtype=int32, numpy=
array([[4, 7, 9],
       [5, 2, 8]], dtype=int32)>

In [290]:
# Vizualizare formei Tensor_0 
Tensor_0.shape

TensorShape([2, 3])

In [291]:
# Redimensionarea tensorului Tensor_0 din (2,3) in (3,2) utizand tf.reshape()
Tensor_3 = tf.reshape(Tensor_0, shape=(3, 2))
Tensor_3

<tf.Tensor: shape=(3, 2), dtype=int32, numpy=
array([[4, 7],
       [9, 5],
       [2, 8]], dtype=int32)>

In [292]:
# Vizualizare formei Tensor_3 
Tensor_3.shape

TensorShape([3, 2])

In [293]:
# Redimensionarea tensorului Tensor_0 utizand tf.transpose()
Tensor_4 = tf.transpose(Tensor_0)
Tensor_4

<tf.Tensor: shape=(3, 2), dtype=int32, numpy=
array([[4, 5],
       [7, 2],
       [9, 8]], dtype=int32)>

In [294]:
# Vizualizare formei Tensor_4 
Tensor_4.shape

TensorShape([3, 2])

In [295]:
# Compararea Tensor_3 si Tensor_4
Tensor_3 == Tensor_4

<tf.Tensor: shape=(3, 2), dtype=bool, numpy=
array([[ True, False],
       [False, False],
       [False,  True]])>

## 3.4. Modificarea tipului de date a unui tensor

In [296]:
# Crearea unui tensor cu elemente float de tip de date implicit (float32)
A = tf.constant([1.7, 7.4])
A.dtype

tf.float32

In [297]:
# Modificarea tipului din float32 in float16 (reducerea preciziei) utilizand tf.cast()
A = tf.cast(A, dtype=tf.float16)
A.dtype

tf.float16

In [298]:
# Crearea unui alt tensor cu elemente intregi de tip de date implicit (int32)
B = tf.constant([9, 4])
B.dtype

tf.int32

In [299]:
# Modificarea tipului din int32 in float32 utilizand tf.cast()
B = tf.cast(B, dtype=tf.float32)
B.dtype

tf.float32

## 3.5. Obtinerea valorilor absolute ale elementelor unui tensor

In [300]:
# Crearea unui tensor cu valori negative
C = tf.constant([-9, -4])
C

<tf.Tensor: shape=(2,), dtype=int32, numpy=array([-9, -4], dtype=int32)>

In [301]:
# Obtinerea valorilor absolute
tf.abs(C)

<tf.Tensor: shape=(2,), dtype=int32, numpy=array([9, 4], dtype=int32)>

## 3.6. Agregarea tensorilor (max, min, mean, sum)

In [302]:
# Crearea unui tensor cu 50 de valori aliatoare in gama 0 - 100
D = tf.constant(np.random.randint(low=0, high=100, size=50))
D

<tf.Tensor: shape=(50,), dtype=int64, numpy=
array([96, 97, 73, 87, 46, 40, 72, 18, 23, 19, 28, 35, 88, 11, 38, 67, 25,
       49, 47, 74, 58, 78, 81, 86,  2, 34, 12, 92, 45, 74, 66, 66, 24, 52,
       52, 43, 27, 76, 77,  5,  7, 69, 27, 43, 61,  0, 31, 29, 92, 69])>

In [303]:
# Determinarea valorii minime a tensorului D utilizand tf.reduce_min
tf.reduce_min(D)

<tf.Tensor: shape=(), dtype=int64, numpy=0>

In [304]:
# Determinarea valorii maxime a tensorului D utilizand tf.reduce_max
tf.reduce_max(D)

<tf.Tensor: shape=(), dtype=int64, numpy=97>

In [305]:
# Determinarea valorii medii a elementelor tensorului D utilizand tf.reduce_mean
tf.reduce_mean(D)

<tf.Tensor: shape=(), dtype=int64, numpy=50>

In [306]:
# Determinarea sumei valorilor elementelor tensorului D utilizand tf.reduce_sum
tf.reduce_sum(D)

<tf.Tensor: shape=(), dtype=int64, numpy=2511>

In [307]:
# Determinarea abaterii medii standart a elementelor tensorului D utilizand tf.math.reduce_std
tf.math.reduce_std(tf.cast(D, tf.float32))

<tf.Tensor: shape=(), dtype=float32, numpy=27.519657>

In [308]:
# Determinarea abaterii medii patratice a elementelor tensorului D utilizand tf.math.reduce_variance
tf.math.reduce_variance(tf.cast(D, tf.float32))

<tf.Tensor: shape=(), dtype=float32, numpy=757.33154>

## 3.7. Determinarea pozitiei valorii maxime si minime a unui tensor

In [309]:
# Determinarea pozitiei valorii maxime a tensorului D utilizand tf.argmax()
tf.argmax(D)

<tf.Tensor: shape=(), dtype=int64, numpy=1>

In [310]:
# Determinarea pozitiei valorii minime a tensorului D utilizand tf.argmin()
tf.argmin(D)

<tf.Tensor: shape=(), dtype=int64, numpy=45>

In [311]:
# Determinarea valorii maxime a tensorului D prin diverse metode
print(f"Valoarea maxima a lui D se afla pe pozitia: {tf.argmax(D).numpy()}") 
print(f"Valoarea maxima utilizand tf.reduce_max(D) este: {tf.reduce_max(D).numpy()}") 
print(f"Valoarea maxima utilizand indexul tf.argmax(D) este: {D[tf.argmax(D)].numpy()}")
print(f"Sunt aceste doua valori identice? {D[tf.argmax(D)].numpy() == tf.reduce_max(D).numpy()}")

Valoarea maxima a lui D se afla pe pozitia: 1
Valoarea maxima utilizand tf.reduce_max(D) este: 97
Valoarea maxima utilizand indexul tf.argmax(D) este: 97
Sunt aceste doua valori identice? True


## 3.8. Stergerea dimensiunilor unitarea

In [312]:
# Crearea unui tensor de 5 dimeniuni cu 50 valori aliate in gama 0 - 100
E = tf.constant(np.random.randint(0, 100, 50), shape=(1, 1, 1, 1, 50))
E

<tf.Tensor: shape=(1, 1, 1, 1, 50), dtype=int64, numpy=
array([[[[[79, 11, 67, 59, 85, 83, 56, 33, 34, 18, 89, 84, 10,  8, 26,
           61, 96, 23, 95, 34, 92, 77, 77, 94, 54, 78, 78, 92, 91,  7,
           16, 27, 26, 98, 83, 10, 75, 60, 12,  3, 67,  1, 56, 36,  2,
           74, 12, 96, 92, 98]]]]])>

In [313]:
 # Vizualizarea formei si dimensiuni tensorului E
E.shape, E.ndim

(TensorShape([1, 1, 1, 1, 50]), 5)

In [314]:
# Stergerea tutror dimensiunilor unitarea a tensorului E utilizand tf.squeeze()
E = tf.squeeze(E)
E

<tf.Tensor: shape=(50,), dtype=int64, numpy=
array([79, 11, 67, 59, 85, 83, 56, 33, 34, 18, 89, 84, 10,  8, 26, 61, 96,
       23, 95, 34, 92, 77, 77, 94, 54, 78, 78, 92, 91,  7, 16, 27, 26, 98,
       83, 10, 75, 60, 12,  3, 67,  1, 56, 36,  2, 74, 12, 96, 92, 98])>

In [315]:
# Vizualizarea formei si dimensiuni noului tensor E
E.shape, E.ndim

(TensorShape([50]), 1)

## 3.9. One-hot encoding

In [316]:
# Crearea unei liste cu indici
index_list = [0, 1, 2, 3]
index_list

[0, 1, 2, 3]

In [317]:
# Transformarea valorilor listei utilizand tf.one_hot()
 # One hot encode them
tf.one_hot(index_list, depth=4)

<tf.Tensor: shape=(4, 4), dtype=float32, numpy=
array([[1., 0., 0., 0.],
       [0., 1., 0., 0.],
       [0., 0., 1., 0.],
       [0., 0., 0., 1.]], dtype=float32)>

In [318]:
# Transformarea valorilor listei modificand parametrul depth al tf.one_hot() 
 # One hot encode them
tf.one_hot(index_list, depth=3)

<tf.Tensor: shape=(4, 3), dtype=float32, numpy=
array([[1., 0., 0.],
       [0., 1., 0.],
       [0., 0., 1.],
       [0., 0., 0.]], dtype=float32)>

In [319]:
# Substituirea valorilor 1 si 0 cu parametri on_value si off_value
tf.one_hot(index_list, depth=4, on_value="Da", off_value="Nu" )

<tf.Tensor: shape=(4, 4), dtype=string, numpy=
array([[b'Da', b'Nu', b'Nu', b'Nu'],
       [b'Nu', b'Da', b'Nu', b'Nu'],
       [b'Nu', b'Nu', b'Da', b'Nu'],
       [b'Nu', b'Nu', b'Nu', b'Da']], dtype=object)>

## 3.10. Alte operatii asupra tensorilor (patratul, log, radacina partrata)

In [320]:
# Crearea unui tensor
F = tf.constant(np.arange(1, 10), dtype=tf.float16)
F

<tf.Tensor: shape=(9,), dtype=float16, numpy=array([1., 2., 3., 4., 5., 6., 7., 8., 9.], dtype=float16)>

In [321]:
# Ridicarea la patrat a valorilor tensorului F
tf.square(F)

<tf.Tensor: shape=(9,), dtype=float16, numpy=array([ 1.,  4.,  9., 16., 25., 36., 49., 64., 81.], dtype=float16)>

In [322]:
# Determinarea radacinii patratice a valorilor tensorului F
tf.sqrt(F)

<tf.Tensor: shape=(9,), dtype=float16, numpy=
array([1.   , 1.414, 1.732, 2.   , 2.236, 2.45 , 2.646, 2.828, 3.   ],
      dtype=float16)>

In [323]:
# Determinarea logaritmului natural al valorilor tensorului F
tf.math.log(F)

<tf.Tensor: shape=(9,), dtype=float16, numpy=
array([0.    , 0.6934, 1.099 , 1.387 , 1.609 , 1.792 , 1.946 , 2.08  ,
       2.197 ], dtype=float16)>

## 3.11. Manipularea tensorilor variabili (tf.Variable)

In [324]:
# Crearea unui tensor variabil
G = tf.Variable(np.arange(0, 5))
G

<tf.Variable 'Variable:0' shape=(5,) dtype=int64, numpy=array([0, 1, 2, 3, 4])>

In [325]:
# Inscrierea valorii 5 pe ultima pozitie
G.assign([0, 1, 2, 3, 5])

<tf.Variable 'UnreadVariable' shape=(5,) dtype=int64, numpy=array([0, 1, 2, 3, 5])>

In [326]:
# Adunarea fiecarui element cu valoarea 10
G.assign_add([10, 10, 10, 10, 10])
G

<tf.Variable 'Variable:0' shape=(5,) dtype=int64, numpy=array([10, 11, 12, 13, 15])>

# 4. Tensorii si NumPy

In [327]:
# Crearea unui tensor direct din numpy array
H = tf.constant(np.array([3., 7., 10.]))
H

<tf.Tensor: shape=(3,), dtype=float64, numpy=array([ 3.,  7., 10.])>

In [328]:
# Conversia unui tensor in NumPy utilizand np.array()
arr = np.array(H)
arr

array([ 3.,  7., 10.])

In [329]:
type(H), type(arr)

(tensorflow.python.framework.ops.EagerTensor, numpy.ndarray)

In [330]:
# Crearea a doi tensori, unu din numpy si altul din tensorflow 
I = tf.constant(np.array([9., 7., 10.]))
J = tf.constant([9., 7., 10.])
I,J

(<tf.Tensor: shape=(3,), dtype=float64, numpy=array([ 9.,  7., 10.])>,
 <tf.Tensor: shape=(3,), dtype=float32, numpy=array([ 9.,  7., 10.], dtype=float32)>)

In [331]:
# Vizualizarea tipului implicit de date a celor doi tensori I si J
I.dtype, J.dtype

(tf.float64, tf.float32)

# 5. Utilizarea @tf.function (cresterea vitezei functiilor Python)

In TensorFlow deseori functiile Python au un decorator @tf.function.

Mai multe despre decorator se poate gasi pe adresa: https://realpython.com/primer-on-python-decorators/

In cazul decoratorului @tf.function, functia Python se transforma intr-o functie ce poate fi apelata de TensorFlow, adica creste viteza de procesare a acesteia. Mai multe despre @tf.function se poate gasi pe adresa: https://www.tensorflow.org/guide/function 


In [332]:
# Crearea unei functii simple in Python
def operatii(x, y):
  return x ** 2 + y

In [333]:
#Crearea a doi tensori
x = tf.constant(np.arange(0, 10))
y = tf.constant(np.arange(10, 20))
x, y

(<tf.Tensor: shape=(10,), dtype=int64, numpy=array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])>,
 <tf.Tensor: shape=(10,), dtype=int64, numpy=array([10, 11, 12, 13, 14, 15, 16, 17, 18, 19])>)

In [334]:
# Apelarea functiei cu trecerea celor doi tensori ca parametri
operatii(x, y)

<tf.Tensor: shape=(10,), dtype=int64, numpy=array([ 10,  12,  16,  22,  30,  40,  52,  66,  82, 100])>

In [335]:
# Crearea aceleiasi functii cu decorarea tf.function
@tf.function
def operatii(x, y):
  return x ** 2 + y

In [336]:
# Apelarea functiei cu trecerea celor doi tensori ca parametri
operatii(x, y)

<tf.Tensor: shape=(10,), dtype=int64, numpy=array([ 10,  12,  16,  22,  30,  40,  52,  66,  82, 100])>

Rezultatele sunt aceleasi deosebirea constand in viteza de operare

# 6. Utilizarea GPU in TensorFlow (sau TPU)

In [337]:
# Verificarea procesorului la care este conectat proiectul
tf.config.list_physical_devices()

[PhysicalDevice(name='/physical_device:CPU:0', device_type='CPU'),
 PhysicalDevice(name='/physical_device:GPU:0', device_type='GPU')]

Proiectul se executa pe CPU (Central Processing Unit). Mai multe pe adresa: https://ro.wikipedia.org/wiki/Unitate_central%C4%83_de_prelucrare

In [338]:
# Verificarea conexiunii la GPU
print(tf.config.list_physical_devices('GPU'))

[PhysicalDevice(name='/physical_device:GPU:0', device_type='GPU')]


Daca se obtine o lista nula "[ ]" atunci proiectul nu se executa pe GPU. Mai multe despre GPU pe adresa: https://ro.wikipedia.org/wiki/Unitate_de_procesare_grafic%C4%83



In Google Colab conectarea la GPU se realizeaza urmand cale din meniu Executie-> Shimbati tipul executie -> Accelerator de hardware GPU  (Runtime -> Change Runtime Type -> Select GPU). Dupa aceasta operatie jupyter notebook se va restarta.

In [341]:
# Repetarea operatie de verificare a conexiunii dupa realizarea operateie de mai sus
print(tf.config.list_physical_devices('GPU'))

[PhysicalDevice(name='/physical_device:GPU:0', device_type='GPU')]


In [342]:
# Informatii despre GPU utilizand !nvidia-smi.
!nvidia-smi

Fri May 21 10:54:44 2021       
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 465.19.01    Driver Version: 460.32.03    CUDA Version: 11.2     |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|                               |                      |               MIG M. |
|===============================+======================+======================|
|   0  Tesla T4            Off  | 00000000:00:04.0 Off |                    0 |
| N/A   65C    P0    28W /  70W |    224MiB / 15109MiB |      0%      Default |
|                               |                      |                  N/A |
+-------------------------------+----------------------+----------------------+
                                                                               
+-------

# 7. Exercitii de exersare

1. Creati un scalar, un vector, o matrice si un tensor cu valori proprii utilizand tf.constant().
2. Determinati parametrii shape, rank si size pentru tensorul creat in punctul 1.
3. Creati doi tensori ce contin valori aliatoare in gama 0 - 1 si au forma [5, 300].
4. Inmultiti cei doi tensori din punctul 3 utilizand tf.matmul().
5. Inmultiti cei doi tensori din punctul 3 utilizand tf.tensordot().
6. Creati un tensor ce contine valori aliatoare in gama 0 - 1 si are forma  [224, 224, 3].
7. Determinati valorile maxime si minime ala tensorului creat in punctul 6.
8. Creati un tensor ce contine valori aliatoare si are forma [1, 224, 224, 3] apoi transformatii forma in [224, 224, 3].
9. Creati un tensor ce contine propriile valori si are forma [10] iar apoi deteminati pozitia pe care se afla valoare maxima si minima.
10. Aplicati One-hot encode asupra tensorului  creat in pinctul 9.